File: F_Sharp/F_Sharp/009.fs
[<AutoOpen>]
module _9

open System

// A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
//
// a2 + b2 = c2
// For example, 32 + 42 = 9 + 16 = 25 = 52.
//
// There exists exactly one Pythagorean triplet for which a + b + c = 1000.
// Find the product abc.

let square n = n * n



let Run() =
    let triplet = seq { for i in 1..1000 do for j in 1..i do yield (i, j) } 
                    |> Seq.map (fun (a, b) -> a, b, (sqrt ((square (float a)) + (square (float b)))))
                    |> Seq.filter (fun (a,b,c) -> round c = c )
                    |> Seq.find (fun (a, b, c) -> a + b + int(c) = 1000)
    
    let a, b, c = triplet

    sprintf "a * b * c = %d * %d * %d = %d" a b (int c) (a * b * int(c))

File: Python/009.py
import sys
import os
import timeit
import math
os.chdir(os.path.dirname(os.path.abspath(__file__))) 

###############################################################################
# A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

# a2 + b2 = c2
# For example, 32 + 42 = 9 + 16 = 25 = 52.

# There exists exactly one Pythagorean triplet for which a + b + c = 1000.
# Find the product abc.
###############################################################################


def find_pythagorean_triplet(n):
	a = 0
	b = 0
	while True:
		a += 1
		if a > n:
			return None

		b = 0
		while b < a:
			b += 1
			c = math.sqrt (a*a + b*b)

			if c.is_integer() and a + b + round(c) == n:
				return (b,a,round(c))

	return None


def main():
	n = 1000
	ret = find_pythagorean_triplet(n)
	if ret != None:
	    (a, b, c) = ret
	    print ("a * b * c = {} * {} * {} = {}".format(a, b, c, a * b * c))
	else:
	    print("Cannot find pythagorean triplet for a + b + c = {}".format(n))


print("Time: {}".format(timeit.timeit('main()', "from __main__ import main", number=1)))


